Enhanced the output viewer to display html outputs (in windows machines), improve viewing of graphs and text outputs and improve processing of objects

Patowhiz opened the pull request:
This PR replaces PR #7782. It aims to omit the CEF packages from GitHub.
@lloyddewit I managed to omit the CEF packages from GitHub. 
I did this by copying changes manually because of the urgency.
The code changes reviewed in PR #7782 are still the same.
@africanmathsinitiative/developers this is ready for review.
Fixes #7121
Fixes #7808
Fixes #7778
Fixes partly #7727

Follow up issues have been outlined in issue #7852.

lloyddewit commented:
@Patowhiz Congratulations on getting rid of the 700+ CEF files.
I checked that I didn't have the CEF packages installed and opened the PR branch. I expected that I'd have to install the CEF NuGet packages manually but I attempted a build anyway. I got the following messages:

![image](https://user-images.githubusercontent.com/57253949/189093242-85292071-293e-43a5-8b71-406a047de7c1.png)

I closed and reopened Visual Studio and did a rebuild solution. I got the following warnings:

![image](https://user-images.githubusercontent.com/57253949/189094126-5d130f23-b7a5-43e3-95ef-b3b73c1e372e.png)

I checked my NuGet packages and Visual Studio had automatically installed the CEF packages:

![image](https://user-images.githubusercontent.com/57253949/189094803-e6ddfbdf-6177-431f-91d2-3c3693eb3b4f.png)

These were older versions so I upgraded 'CefSharp.WinForms' to the latest version. This automatically upgraded the 3 dependency packages:

![image](https://user-images.githubusercontent.com/57253949/189095879-af86d458-c9a9-4153-a492-07d509547634.png)

But I still got the same three warnings. Do you get these warnings also? Can we remove them?

I will now start the code peer review.

Patowhiz commented:
@lloyddewit I think that's just because the dll targets an X86 CPU architecture.  That's what the message says.
CefSharp typically comes with both X64 and X86 dlls. Then during runtime, loads the appropriate dll by detecting the set architecture from the program assembly file.

lloyddewit commented:
> @lloyddewit I think that's just because the dll targets an X86 CPU architecture. That's what the message says. CefSharp typically comes with both X64 and X86 dlls. Then during runtime, loads the appropriate dll by detecting the set architecture from the program assembly file.

yes, I think you're correct - thank you

lloyddewit commented on instat/frmMaximiseOutput.vb:
```diff
@@ -0,0 +1,81 @@
+                If Path.GetExtension(_strDisplayedFileName).ToLower = ".html" Then
+                    'copy the html file with it's associated directory; css, javascript files etc
+                    For Each foundDirectory As String In My.Computer.FileSystem.GetDirectories(
+                        Path.GetDirectoryName(_strDisplayedFileName),
+                        FileIO.SearchOption.SearchTopLevelOnly,
+                        "*" & Path.GetFileNameWithoutExtension(_strDisplayedFileName) & "*")
```
It's a bit confusing which lines are part of the `For` condition. Indent o make clearer?
```suggestion
                            Path.GetDirectoryName(_strDisplayedFileName),
                            FileIO.SearchOption.SearchTopLevelOnly,
                            "*" & Path.GetFileNameWithoutExtension(_strDisplayedFileName) & "*")
```

lloyddewit commented on instat/ucrInputComboBox.vb:
```diff
@@ -11,359 +11,337 @@
+Imports System.ComponentModel
+
+Public Class ucrInputComboBox
+    Dim _strRObjectItemsTypeLabel As String = ""
```
```suggestion
    Private _strRObjectItemsTypeLabel As String = ""
```

lloyddewit commented on instat/ucrInputComboBox.vb:
```diff
@@ -11,359 +11,337 @@
+    End Sub
+
+    Public Sub SetRObjectItemsTypeLabel(strRObjectLabel As String)
+        Me._strRObjectItemsTypeLabel = strRObjectLabel
```
```suggestion
        _strRObjectItemsTypeLabel = strRObjectLabel
```

lloyddewit commented on instat/ucrInputComboBox.vb:
```diff
@@ -11,359 +11,337 @@
+    End Sub
+
+    Private Sub FillItemTypes()
+        Select Case _strRObjectItemsTypeLabel
```
If we do this check before the `Select`, then we can remove all the `                If ucrDataFrameSelector IsNot Nothing Then` checks below.
```suggestion
        If ucrDataFrameSelector Is Nothing Then
            Exit Sub
        End If
        Select Case _strRObjectItemsTypeLabel
```

lloyddewit commented on instat/ucrSave.vb:
```diff
@@ -650,14 +650,27 @@ Public Class ucrSave
+                            Case RObjectTypeLabel.Graph,
+                                 RObjectTypeLabel.Summary,
+                                 RObjectTypeLabel.Table
+                                If Me._strRObjectLabel = RObjectTypeLabel.Table AndAlso String.IsNullOrEmpty(Me._strRObjectFormat) Then
```
```suggestion
                                If _strRObjectLabel = RObjectTypeLabel.Table AndAlso String.IsNullOrEmpty(_strRObjectFormat) Then
```

lloyddewit reviewed (changes requested):
@Patowhiz Thank you for this large and significant PR.
I only have fairly minor comments

Patowhiz commented:
@rdstern @lloyddewit  this is now ready for you.

@rdstern As discussed I've added the new implementation of setting the heights of the outputs. The setting is in the options dialog. The English description of this new setting  is not available because it is being translated. I think the translation bit can be done in a separate PR. I've set maximum output heights to be 300 by default.

Also not, objects are not being displayed on the view objects dialog. I noticed I have to make changes on different controls for this. I think it's best to do that in a different PR, I now have this task as my top priority.

lloyddewit commented:
@Patowhiz there are 15-20 open comments above. If you resolve them and let me know when done, then I'll review this PR again.
When you resolve the comments, please do not click 'Resolve Conversation' (otherwise I cannot keep track of which comments I've reviewed). I will click 'Resolve Conversation' next time I review.
Thanks!

Patowhiz commented on instat/Model/Output/clsOutputElement.vb:
```diff
@@ -13,39 +13,38 @@
-        _lstStringOutput = New List(Of String)
-        _lstBmpImage = New List(Of Bitmap)
     End Sub
-
```
I have removed the empty lines.

Patowhiz commented on instat/Model/Output/clsOutputElement.vb:
```diff
@@ -58,17 +57,25 @@ Public Class clsOutputElement
+    End Property
+    Public ReadOnly Property HtmlOutput As String
+        Get
+            Return _lstHtmlOutput.FirstOrDefault
```
Done.

Patowhiz commented on instat/Model/Output/clsOutputElement.vb:
```diff
@@ -88,21 +94,29 @@ Public Class clsOutputElement
     End Property
-
     Public Function Clone() As clsOutputElement
         Return Me.MemberwiseClone
```
It has 2 references. Should we refactor the possible issue in a separate PR?

lloyddewit replied:
> It has 2 references

I couldn't see these, am I missing something?

> Should we refactor the possible issue in a separate PR?

It should be fairly straightforward to explicitly clone each list in this class. I think it's just 5 lines of code.

lloyddewit replied:
@Patowhiz thank you for the screenshot. It's strange I don't see the references. Maybe it's connected to the compilation errors I get in VS.

Regarding the cloning, I origionally thought it would be enough to clone the 5 lists manually but of course, this would create new lists but the list elements would still contain shared references. We may need to clone on multiple levels.
So I agree that a todo is enough for now.

Patowhiz commented on instat/Model/Output/clsOutputElement.vb:
```diff
@@ -88,21 +94,29 @@ Public Class clsOutputElement
     End Function
-
+    Public Sub AddTextOutput(strFileName As String, script As List(Of clsRScriptElement))
+        _lstTextOutput.Add(strFileName)
```
I discussed this with @ChrisMarsh82 .
He anticipated that in future, an output element could have multiple outputs attached to it.
I can see how that can be implemented differently but I didn't want to spend a lot of time on this because he is already working on different output structures that deprecate what we have.

lloyddewit replied:
ok, understood

Patowhiz commented on instat/Model/Output/clsOutputLogger.vb:
```diff
@@ -69,27 +64,80 @@ Public Class clsOutputLogger
+            Case ".txt"
+                AddTextOutput(strFileName)
+            Case Else
+                'todo. developer error
```
Done.

Patowhiz commented on instat/Model/Output/clsOutputLogger.vb:
```diff
@@ -184,4 +211,10 @@ Public Class clsOutputLogger
-End Class
\ No newline at end of file
+End Class
+
+
+
+
+
+
```
Removed.

Patowhiz commented on instat/UserControl/ucrOutputPage.vb:
```diff
@@ -217,13 +219,24 @@ Public Class ucrOutputPage
+                Clipboard.SetImage(GetBitmapFromFile(element.ImageOutput))
                 richText.Paste()
+            Case OutputType.HtmlOutput
+                'todo
```
Actually, this subroutine needs to be refactored. The image and html will never be inside an rtf control. they now have their own controls. We should do the same for Script and Text. for script, it could be the Scintilla control that we have in the output window.
I suggest we do that in separate PR.

lloyddewit replied:
ok

Patowhiz commented on instat/UserControls/Webview/Windows/CefRuntimeWrapper.vb:
```diff
@@ -0,0 +1,111 @@
+    Private Sub New()
+    End Sub
+
+    Public Shared Function InitialiseCefRuntime() As Boolean
```
Done

Patowhiz commented on instat/UserControls/Webview/Windows/ucrWebViewer.vb:
```diff
@@ -0,0 +1,87 @@
+'
+' You should have received a copy of the GNU General Public License
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
```
Done.

Patowhiz commented on instat/UserControls/ucrTextViewer.vb:
```diff
@@ -0,0 +1,94 @@
+﻿Imports System.IO
+Imports System.Runtime.InteropServices
+
+Public Class ucrTextViewer
```
Done.

Patowhiz commented on instat/UserControls/ucrTextViewer.vb:
```diff
@@ -0,0 +1,94 @@
+Public Class ucrTextViewer
+    Inherits Panel
+
+    Public ReadOnly _richTextBox As RichTextBox
```
In a separate PR. I do intend to have this as private, after finding a work around in regards to output window scrolling bug. in the meantime I have changed it as suggested.

Patowhiz commented on instat/UserControls/ucrTextViewer.vb:
```diff
@@ -0,0 +1,94 @@
+
+    Public Sub AddText(text As String)
+        _richTextBox.AppendText(text)
+        'by default always set the height of this control to correspond to the heigh of the rich text box contents
```
Done.

Patowhiz commented on instat/frmMain.vb:
```diff
@@ -82,6 +83,9 @@ Public Class frmMain
         clsOutputLogger = New clsOutputLogger
         clsRLink = New RLink(clsOutputLogger)
+        If RuntimeInformation.IsOSPlatform(OSPlatform.Windows) Then
+            CefRuntimeWrapper.InitialiseCefRuntime()
```
Done. A message will be shown to the user. I'm to get suggestions on a better message.

Patowhiz commented:
@lloyddewit I have addressed all the comments.
@rdstern kindly retest this.
Before testing please note the following;

**Dialogs that demonstrate the new changes**
1. Prepare > Check Data > One Variable Summaries. Demonstrates how text and table outputs are displayed. 
2. Prepare > Check Data > View/Delete Labels.
3. Describe > Two/Three variables > Pivot Table. Demonstrates complex html outputs.
4. Describe > Specific Tables/Graphs > Frequency / Summary Tables.
5. Describe > Specific Tables/Graphs > Point (Scatter) Plot.

**Dialogs that need discussions before refactoring**
1. Describe > Two/Three variables > Summarise

**Dialogs that can be refactored by the team to use the new implementation**
1. Describe > Two/Three variables > Two-Way Frequencies
2.  Describe > Two/Three variables > Three-Way frequencies
3. All dialogs that have a save control

**Issues known**
1. Every time a html output is produced, the output view adds the web control which initiates a new browser process similar to the way a new tab is created in Chrome. This means more memory usage. There may be need to prioritise how the output viewer can reclaim back that memory when the user is no longer viewing the html output..
2. Currently viewing of html objects that produce 2 html files is not supported. The R function added only supports viewing of a single html file.
4. Plotly html objects may produce html files that don't return the correct CSS height. I'll have to investigate a work around.
6. Some dialogs may take some time to produce the html objects. This is not related to the new control that only reads from the produced html file dynamically.
7. At times the output window does not scroll smoothly. This is an existing issue.
8. Viewing of objects dialog no longer works. Intentional changes to R-Instat object metadata is the cause.
9. Backward incompatibility of rds files saved by previous R-Instat versions. This is only an issue for the very very old versions. Saving of objects was already disabled in many previous versions.

The above issues and others have been outlined in issue #7852 as follow up issues.

**New enhancements and concepts introduced by this PR**
1. Viewing of html objects in R-Instat.
2. Retrieval of R outputs into the .Net front end.
3. Maximising of outputs in the output window.
4. Setting the maximum heights of outputs in the output window.
5. New way of processing and managing objects.
6. New object types. All outputs that go the output window can now be saved as objects. @rdstern this is very key and may affect how we conceptualise a dialog that has an output.
7. Lot's of code enhancements both at R and .Net level.

@rdstern the changes made in this PR affect almost all dialogs in the background. We are aiming for another release before December. I suggest while we wait for @lloyddewit code review, you do all your testing through changes in this PR. I have updated it with the latest changes from the master. I'm keen on identifying any regression I may have introduced early enough.

Thanks.

rdstern reviewed (dismissed):
@Patowhiz this looks great.
I tried all the dialogues you said have changed.  They all seem ok to me, though I couldn't see what was different in the scatterplot (which you said was changed) and another graph.  I tried the bar chart and it seems to behave similarly.  I like that you have still left the R-viewer available, and also tried plotly.

You mention that the Describe > 2/3 variable > Summarise still needs discussion.  Just to add that this dialogue also has a pull request, (#7920) so perhaps that should be merged first.  Vitalis is doing that.  Or perhaps he could include the changes for the new output within it?

Can this be merged now if @lloyddewit is also happy, or do things on objects need to change first?

Patowhiz commented:
@rdstern I'd suggest PR #7920 by @Vitalis95  to wait . The changes made here are meant to demonstrate how dialogs should use the save control.
The scatter plot dialog has a similar change as an example for other graph dialogs.
Of course,  I had to support existing dialogs to prevent regression. 

@rdstern are you happy with the height set for the outputs, they have fixed heights.

We now have a control for editing and viewing scripts. Should I use the control in the output viewer as well?

rdstern commented:
@Patowhiz I comment as I try different things.
I started with an exercise from Tutorial 1 - so the diamonds dataset from ggplot2 in the library.
Using Describe > 2/3 variables > Graph
![image](https://user-images.githubusercontent.com/11226469/203393962-61a95e57-1518-465d-b5ca-19eedde195cd.png)
Here is the output window:
![image](https://user-images.githubusercontent.com/11226469/203394174-5c3fbd21-acbb-4f54-b989-7dcf68d15536.png)

You only see half the graph.  I originally thought it wasn't scrolling to the bottom, but the half is all there is.  You can maximise and that works well. 
But in the output window you only have half the graph as shown more clearly below.
![image](https://user-images.githubusercontent.com/11226469/203394593-2a88cf68-35e8-4a26-9651-26f362b51678.png)

I am now trying more things!

rdstern commented:
@Patowhiz and @ChrisMarsh82 I note that none of the points above prevent the new system being used.  So I wonder about even merging this pull request in its current form?
a) Then there are a few more additions needed, that could proceed now - particularly with other dialogues that currently use a browser.  They could be adapted this week.
b) This could be in parallel with tweaks to this facility.

I remain very keen to have a new version (0.7.8?) in time for an important climatic workshop that starts on 5th December in Reading.  But it is for a group who would be ok (even happy) to be the first on a provisional release.  So it needs to be downloadable from slack, but not necessarily from the website.

Patowhiz commented:
@rdstern thanks for going through the PR. 

Interesting to see why the graph produced is not scrollable. I'll check on that, I'm keen also to find out of there is other graphs produced that are not scrollable.

Remember, we deliberately decided to limit the size of the outputs in the output window, reason being, at times we have very long outputs and they become problematic when scrolling, we didn't have a work around until now that we have the new maximise feature.

I'm aware of the blank black. It relates to rendering in windows 11 machines that have high resolutions, it's not present in windows 10 machine. I'm already thinking about this.

lloyddewit commented on instat/Enum/RObjectTypeLabel.vb:
```diff
@@ -0,0 +1,17 @@
+    Public Const Graph As String = "graph"
+    Public Const Table As String = "table"
+    Public Const Model As String = "model"
+    Public Const Structure_label As String = "structure"
```
not yet resolved

lloyddewit commented on instat/clsRLink.vb:
```diff
@@ -768,6 +701,48 @@ Public Class RLink
+
     '''--------------------------------------------------------------------------------------------
     ''' <summary>
     ''' This method executes the <paramref name="strScript"/> R script and displays the output. The
```
not yet done

Patowhiz replied:
Done.

lloyddewit commented on instat/clsRLink.vb:
```diff
@@ -813,11 +788,11 @@ Public Class RLink
     ''' displays the exception message.</param>
     '''--------------------------------------------------------------------------------------------
-    Public Sub RunScript(strScript As String, Optional iCallType As Integer = 0, Optional strComment As String = "", Optional bSeparateThread As Boolean = True, Optional bShowWaitDialogOverride As Nullable(Of Boolean) = Nothing, Optional bUpdateGrids As Boolean = True, Optional bSilent As Boolean = False)
+    Public Sub RunScript(strScript As String, Optional iCallType As Integer = 0, Optional bAddOutputInLog As Boolean = True, Optional strComment As String = "", Optional bSeparateThread As Boolean = True, Optional bShowWaitDialogOverride As Nullable(Of Boolean) = Nothing, Optional bUpdateGrids As Boolean = True, Optional bSilent As Boolean = False)
```
not yet done.
Also, there are some other inconsistencies in the header parameter list. Correct these at the same time?

Patowhiz replied:
Done. It may change to the a string in a follow up PR, especially once we deprecate the iCallType parameter.

lloyddewit commented:
@Patowhiz Thank you for all the changes. I resolved all the open comments apart from 5. Please see above.

I also have many compilation errors (see example below). It seems to be related to VB and WinForms classes. Did something change in the project configuration? If you don't have this problem, then should we ask another developer to test in case the problem is with my system?

If we can resolve the above, then I'm happy to approve.

Thanks

![image](https://user-images.githubusercontent.com/57253949/203526495-0b6717c7-b3c0-4e1b-b5ae-dd2dc04cf800.png)

Patowhiz commented:
@lloyddewit this is odd. It compiles and runs fine in 2 machines. @rdstern did you get the same compilation errors, I presume you were able to test because it compiled fine on your end.
@N-thony could you try testing this PR in your machine

rdstern commented:
@Patowhiz and @lloyddewit it ran on my machine.  And I am usually the odd one out here - that it doesn't run on my machine, but is ok on many others.
By the way you mentioned (I think) that some of my problems above were ok in Windows 10.  I am still using Windows 10.  I haven't changed to Windows 11 yet.

N-thony commented:
> @lloyddewit this is odd. It compiles and runs fine in 2 machines. @rdstern did you get the same compilation errors, I presume you were able to test because it compiled fine on your end. @N-thony could you try testing this PR in your machine

@Patowhiz  @lloyddewit I have tried with my machine, this is what I got
![image](https://user-images.githubusercontent.com/68591383/203532158-a628d04c-d094-4ac0-a575-8449210fd96a.png)

Patowhiz commented:
@rdstern `Describe > 2/3 Variables > 3-way frequencies` dialog will have to be refactored to use the "new system". That can be done by the team. As explained in my comments above and in issue #7852 any dialog that produces a html output will have to use the "new system"

@N-thony `Climatic > Check Data > Inventory` plot dialog is implemented incorrected. I think the R function used as a `before code` should be used as the `base code`. It currently has no base function. That's why @rdstern got the error above, which I have now made sure that dialogs implemented in such incorrect ways won't raise that error. I still suggest the dialog to be correctly implemented. Perhaps @anastasia-mbithe could do it or any other team member you suggest. Thanks

Patowhiz commented:
@N-thony thanks for testing whether this PR compiles.
@lloyddewit my machine, @rdstern and @N-thony are able to compile the changes here.  The Github build checks are also successful. 
I wonder what could be happening in yours.

rdstern reviewed (approved):
@Patowhiz and @lloyddewit I am still getting the same features as reported earlier.  But no problem compiling etc.  
I am approving, as I can live with those problems for now, and it will be good for the team to work on the remaining dialogues that currently give html output in a browser.

lloyddewit commented:
@Patowhiz Regarding my compilation errors, I deleted the PR branch, made a new PR branch and did a 'rebuild solution'. It automatically downloaded the CefSharp library, fixed the compilation errors and built successfully.
Happy to merge.
Thank you for this big piece of work!

Patowhiz commented:
Thanks @lloyddewit.
